networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandler.java

```diff
@@ -105,24 +105,45 @@ public void onIncomingMessage(
     final UInt64 count = maxRequestSize.min(message.getCount());
     final UInt64 endSlot = message.getStartSlot().plus(message.getStep().times(count)).minus(ONE);
 
-    final UInt64 headBlockSlot =
-        combinedChainDataClient.getBestBlock().map(SignedBeaconBlock::getSlot).orElse(ZERO);
-    final NavigableMap<UInt64, Bytes32> hotRoots;
-    if (combinedChainDataClient.isFinalized(endSlot)) {
-      // All blocks are finalized so skip scanning the protoarray
-      hotRoots = new TreeMap<>();
-    } else {
-      hotRoots =
-          combinedChainDataClient.getAncestorRoots(
-              message.getStartSlot(), message.getStep(), count);
-    }
-    // Don't send anything past the last slot found in protoarray to ensure blocks are consistent
-    // If we didn't find any blocks in protoarray, every block in the range must be finalized
-    // so we don't need to worry about inconsistent blocks
-    final UInt64 headSlot = hotRoots.isEmpty() ? headBlockSlot : hotRoots.lastKey();
-    return sendNextBlock(
-        new RequestState(
-            message.getStartSlot(), message.getStep(), count, headSlot, hotRoots, callback));
+    return combinedChainDataClient
+        .getEarliestHistoricalBlockSlot()
+        .thenCompose(
+            earliestSlot -> {
+              if (earliestSlot.map(s -> s.isGreaterThan(message.getStartSlot())).orElse(true)) {
```

Comment:
Just to make sure: are we neglecting the case when we might have some blocks matching the request? I mean the case where `earliestSlot < endSlot`

Reply:
Yeah, that's right.  If we return only later blocks within the range, the response could be interpreted to mean that the earlier slots are skipped blocks, when we don't know that.